### DISCUSSION TITLE
semanticTokens don't seem to generate in nx/monorepo/storybook context

### DISCUSSION THREAD

**Discussion by mayteio** [chakra-ui/panda#988] [General]
### Description

I'm using `nx` to manage my project with an "integrated" setup (i.e. using @nx/next). I've got it set up so that;
1. Main Next.js app runs in `apps/my-app`
2. tokens, semantic tokens and recipes live in `libs/theme` and imported into the `apps/my-app/panda.config.ts` via `@nx-integrated-monorepo/theme`.
3. panda watches the next.js app/storybook and outputs are sent to `libs/styled-system/src/__generated__/*` and exported via `@nx-integrated-monorepo/styled-system`.

This works as expected for recipes and tokens, but not semantic tokens.

Here's my panda.config.ts:

```ts
import { defineConfig } from '@pandacss/dev';
import {
  recipes,
  semanticTokens,
  tokens,
  textStyles,
} from '@nx-integrated-monorepo/theme';

export default defineConfig({
  include: ['./app/**/*.{ts,tsx}', '../../libs/**/*.{ts,tsx}'],
  exclude: [],
  theme: {
    semanticTokens, // doesn't work
    extend: {
      recipes,
      textStyles,
      tokens,
      semanticTokens: { // also doesn't work
        colors: {
          heavy: { value: '#FF0000' },
        },
      },
    },
  },
  jsxFramework: 'react',
  outExtension: 'js', // needed for jest
  outdir: '../../libs/styled-system/src/__generated__',
});
```

This mostly works as expected - I can see **tokens** I've created via `defineTokens` in the output CSS, but the semantic tokens are not outputting (regardless of whether I inline them here or define them in the theme lib). 

### Link to Reproduction

[Codesandbox](https://codesandbox.io/p/github/mayteio/panda-nx-next/draft/trusting-ganguly?file=%2Ftsconfig.base.json%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljs6y7t400ka3b6nhjfj5epl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljs6y7t400kc3b6ns2io6ppo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljs6y7t400ka3b6nhjfj5epl%2522%253A%257B%2522id%2522%253A%2522cljs6y7t400ka3b6nhjfj5epl%2522%252C%2522activeTabId%2522%253A%2522cljtcr68s02lq3b6ntx6c60pp%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapps%252Fmy-app%252Fapp%252Fglobal.css%2522%252C%2522id%2522%253A%2522cljsb4ppl00jq3b6m5es96q13%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Ftokens.ts%2522%252C%2522id%2522%253A%2522cljsb5pc100rs3b6m5v4t2lgi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Fbutton.ts%2522%252C%2522id%2522%253A%2522cljsb5ufs00z73b6mr7oqjsjc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Fforms%252Fsrc%252Fbutton%252Findex.tsx%2522%252C%2522id%2522%253A%2522cljsb8fsj01ub3b6mn6038i8k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cljtcdy4m000b3b6ne1v7ermx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Ftsconfig.base.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Ftext-styles.ts%2522%252C%2522id%2522%253A%2522cljtcolg202dp3b6nxmbfkd14%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapps%252Fmy-app%252Fpanda.config.ts%2522%252C%2522id%2522%253A%2522cljtcr68s02lq3b6ntx6c60pp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Findex.ts%2522%252C%2522id%2522%253A%2522cljtcrny802qj3b6ns2dozsfh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522cljte325e07dy3b6nmvd1662w%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522cljs6y7t400kc3b6ns2io6ppo%2522%253A%257B%2522id%2522%253A%2522cljs6y7t400kc3b6ns2io6ppo%2522%252C%2522activeTabId%2522%253A%2522cljte61kp08mh3b6nnz9ce5gg%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljsb0jjj00fn3b6n1u4g1car%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljsb39ji00swe6e0bhjlampu%2522%257D%252C%257B%2522type%2522%253A%2522EMPTY_DEVTOOL%2522%252C%2522id%2522%253A%2522cljte5igh081k3b6negufpscl%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522id%2522%253A%2522cljte5n2508bs3b6n8pg9875r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522port%2522%253A4400%252C%2522id%2522%253A%2522cljte61kp08mh3b6nnz9ce5gg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%253Fpath%253D%252Fstory%252Fapp-page--primary%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce

1. [Go to codesandbox](https://codesandbox.io/p/github/mayteio/panda-nx-next/draft/trusting-ganguly?file=%2Ftsconfig.base.json%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljs6y7t400ka3b6nhjfj5epl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljs6y7t400kc3b6ns2io6ppo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljs6y7t400ka3b6nhjfj5epl%2522%253A%257B%2522id%2522%253A%2522cljs6y7t400ka3b6nhjfj5epl%2522%252C%2522activeTabId%2522%253A%2522cljtcr68s02lq3b6ntx6c60pp%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapps%252Fmy-app%252Fapp%252Fglobal.css%2522%252C%2522id%2522%253A%2522cljsb4ppl00jq3b6m5es96q13%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Ftokens.ts%2522%252C%2522id%2522%253A%2522cljsb5pc100rs3b6m5v4t2lgi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Fbutton.ts%2522%252C%2522id%2522%253A%2522cljsb5ufs00z73b6mr7oqjsjc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Fforms%252Fsrc%252Fbutton%252Findex.tsx%2522%252C%2522id%2522%253A%2522cljsb8fsj01ub3b6mn6038i8k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cljtcdy4m000b3b6ne1v7ermx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Ftsconfig.base.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Ftext-styles.ts%2522%252C%2522id%2522%253A%2522cljtcolg202dp3b6nxmbfkd14%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapps%252Fmy-app%252Fpanda.config.ts%2522%252C%2522id%2522%253A%2522cljtcr68s02lq3b6ntx6c60pp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Flibs%252Ftheme%252Fsrc%252Findex.ts%2522%252C%2522id%2522%253A%2522cljtcrny802qj3b6ns2dozsfh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522cljte325e07dy3b6nmvd1662w%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522cljs6y7t400kc3b6ns2io6ppo%2522%253A%257B%2522id%2522%253A%2522cljs6y7t400kc3b6ns2io6ppo%2522%252C%2522activeTabId%2522%253A%2522cljte61kp08mh3b6nnz9ce5gg%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljsb0jjj00fn3b6n1u4g1car%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljsb39ji00swe6e0bhjlampu%2522%257D%252C%257B%2522type%2522%253A%2522EMPTY_DEVTOOL%2522%252C%2522id%2522%253A%2522cljte5igh081k3b6negufpscl%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522id%2522%253A%2522cljte5n2508bs3b6n8pg9875r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522port%2522%253A4400%252C%2522id%2522%253A%2522cljte61kp08mh3b6nnz9ce5gg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%253Fpath%253D%252Fstory%252Fapp-page--primary%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
2. Run task called "Run storybook"
3. Check `libs/styled-system/src/__generated__tokens/index.css` for missing `heavy` semantic token
4. Verify in storybook that the token isn't found:

<img width="998" alt="Screenshot 2023-07-08 at 12 35 31 pm" src="https://github.com/chakra-ui/panda/assets/43975092/5960d16a-7b46-4dee-9f73-3318a445d855">

### JS Framework

React, TS, next.js, storybook, nx (🥲)

### Panda CSS Version

0.6.0

### Operating System

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by mayteio** [comment]
Hey @segunadebayo, so I know for future - what should be a bug and what should be a discussion?

FWIW I've gone a bit further and it appears the same issue happens for `defineTextStyles` too. It works as expected when running next.js:
<img width="772" alt="Screenshot 2023-07-09 at 4 22 12 pm" src="https://github.com/chakra-ui/panda/assets/43975092/081dd2f6-3bfb-4921-9ebf-b14f0909a953">

But is missing *some* styles when run in storybook (`semanticTokens` and `textStyles` as far as I can tell):
<img width="749" alt="Screenshot 2023-07-09 at 4 22 21 pm" src="https://github.com/chakra-ui/panda/assets/43975092/0c77a703-422f-4c89-91a3-b18dadb6cfe0">

[Updated codesandbox](https://codesandbox.io/p/github/mayteio/panda-nx-next/harley/crazy-banana?file=%2Fapps%2Fmy-app%2Fpanda.config.ts%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljv1xehs000c3b6ntkkz2zds%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljv1xehs000e3b6nkdyl18ir%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljv1xehs000c3b6ntkkz2zds%2522%253A%257B%2522id%2522%253A%2522cljv1xehs000c3b6ntkkz2zds%2522%252C%2522activeTabId%2522%253A%2522cljv1zjho01im3b6nhzw0opf3%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljv1xehr000b3b6nmpdbr7hu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapps%252Fmy-app%252Fpanda.config.ts%2522%257D%252C%257B%2522type%2522%253A%2522DIFF%2522%252C%2522filepath%2522%253A%2522%252Fpnpm-lock.yaml%2522%252C%2522base%2522%253A%2522HEAD%2522%252C%2522id%2522%253A%2522cljv1zjho01im3b6nhzw0opf3%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522cljv1xehs000e3b6nkdyl18ir%2522%253A%257B%2522id%2522%253A%2522cljv1xehs000e3b6nkdyl18ir%2522%252C%2522activeTabId%2522%253A%2522cljv1za5z01fq3b6n5i4lxxps%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520app%2522%252C%2522id%2522%253A%2522cljv1xqii00ag3b6nrw6k5sth%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Run%2520app%2522%252C%2522port%2522%253A4200%252C%2522id%2522%253A%2522cljv1xye400rp3b6nxbxhr4o8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522id%2522%253A%2522cljv1yt5001563b6nir4d81ti%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522port%2522%253A4400%252C%2522id%2522%253A%2522cljv1za5z01fq3b6n5i4lxxps%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%253Fpath%253D%252Fstory%252Fapp-page--primary%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by mayteio** [comment]
FWIW they work in a regular next/storybook context:
<img width="772" alt="Screenshot 2023-07-09 at 4 22 12 pm" src="https://github.com/chakra-ui/panda/assets/43975092/50c8a609-6a21-4a96-a5aa-68a09ac6a16a">
[Codesandbox without nx](https://codesandbox.io/p/github/mayteio/panda-next-storybook/main?layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljvy0vm2000b3b6nk4x7it5m%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljvy0vm2000d3b6nk6tdt1hf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B44.997562224344136%252C55.002437775655864%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljvy0vm2000b3b6nk4x7it5m%2522%253A%257B%2522id%2522%253A%2522cljvy0vm2000b3b6nk4x7it5m%2522%252C%2522activeTabId%2522%253A%2522cljvy5u8g01wp3b6np0o3wak8%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljvy0vm1000a3b6n674l36ia%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522cljvy1s0u00hc3b6ni3th0hr6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapp%252Fpage.tsx%2522%252C%2522id%2522%253A%2522cljvy5u8g01wp3b6np0o3wak8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cljvyhozj03jx3b6nsgxuh7f9%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cljvy0vm2000d3b6nk6tdt1hf%2522%253A%257B%2522id%2522%253A%2522cljvy0vm2000d3b6nk6tdt1hf%2522%252C%2522activeTabId%2522%253A%2522cljvy3m3m01hj3b6ndvqlmolm%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljvy0vm2000c3b6nemyly30v%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522PROJECT_SETUP%2522%257D%252C%257B%2522type%2522%253A%2522EMPTY_DEVTOOL%2522%252C%2522id%2522%253A%2522cljvy2ouz00qe3b6n5so6sph8%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520app%2522%252C%2522id%2522%253A%2522cljvy2qmc00u43b6nmdx7ec6a%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522id%2522%253A%2522cljvy322w013k3b6nwkd6ovxi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%257D%252C%257B%2522id%2522%253A%2522cljvy3gur01cu3b6ntaq1bdjg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A3000%252C%2522taskId%2522%253A%2522Run%2520app%2522%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522id%2522%253A%2522cljvy3m3m01hj3b6ndvqlmolm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A6006%252C%2522taskId%2522%253A%2522Run%2520storybook%2522%252C%2522path%2522%253A%2522%252F%253Fpath%253D%252Fstory%252Fapp-page--primary%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by mayteio** [comment]
Alright, I've been able to solve this thanks to @thetiby here: https://github.com/chakra-ui/panda/discussions/943#discussioncomment-6411631

1. `panda.config.ts` lives in `lib/styled-system` and looks like this:
```ts
export default defineConfig({
  include:
    process.env.NX_TASK_TARGET_TARGET === 'storybook'
      ? ['./apps/**/*.{ts,tsx}', './libs/**/*.{ts,tsx}']
      : ['../../apps/**/*.{ts,tsx}', '../../libs/**/*.{ts,tsx}'],
  outdir:
    process.env.NX_TASK_TARGET_TARGET === 'storybook'
      ? './libs/styled-system/src/__generated__'
      : '../../libs/styled-system/src/__generated__',

  // modify this to add to the theme
  theme: { ... },
  ...
});
```

2. `postcss.config.cjs` lives in `apps/my-app/postcss.config.cjs` and looks like this:
```ts
module.exports = {
  plugins: {
    '@pandacss/postcss': {
      configPath:
        process.env.NX_TASK_TARGET_TARGET === 'storybook'
          ? './libs/styled-system/panda.config.ts'
          : '../../libs/styled-system/panda.config.ts',
    },
  },
};
```

3. `pnpm prepare` in `package.json` looks like this:
```
    "prepare": "panda codegen --cwd=apps/my-app --config ../../libs/styled-system/panda.config.ts"
```

Note: storybook and next.js seem to run with a different working directory in `nx`. Due to this, I had to switch out various config options depending on the task being run.

**Reply by FranciscoKloganB** [reply]
@mayteio Can you explain your solution for dummies? I am new to NX and I really can't understand what's going on.

I was trying not to use native Tailwind for this project I am about to start, and decided to just use the `emotion` recipes provided with Nx since I was already familiar with emotion without having to worry about set up.

Then I heard about Panda CSS and I really wanted to try it out, since it provides the good parts of Tailwind, is React Server Component proof and can generate Framework agnostic styles, which is nice for a mono repo. However, since I do not really understand Nx and I do not know how to create my own generators, I am having trouble understanding what's going on and what would be considered an anti-pattern.

Aren't we supposed to respect module boundaries? If we are copying the output on `prepare` using codegen, aren't we going against this principle?

**edit:** if it would not be asking to much, could you repost your reproduction including the final solution?

**Reply by mayteio** [reply]
Sure thing! The way we've got it set up is with an @nx/next integrated repository (more on [integrated repositories](https://nx.dev/getting-started/tutorials/integrated-repo-tutorial)) - that is there's one package.json for the entire project and the `libs` folder is more of a code organization heuristic. Nx provides tooling for generating/moving code around as well as running tasks (`nx affected` is the point of this - big repositories with lots of tests take a long time in CI/CD, `nx affected` only runs tests/lints in "projects" (think apps or libs) based on code that has changed since `main`.

On to the module boundaries - you can do this (probably in both a package and integrated repo) with panda. How we've got it set up is keeping all panda-related code mostly in a `lib`. We've got a `lib` called `styled-system` that includes the panda config, theme tokens. The config is set to generate inside a `__generated__` folder within the `styled-system` lib. When we run prepare, it generates the code into here. The `postcss.config.cjs` file sits in `apps/your-app` so that it can hook into the next.js and storybook compile steps and re-generate when we make changes there. Since the config knows to look both in app and in our libs, it generates whenever we make changes to any code in our project.

`@your-repo/styled-system` exports the things we need from the `__generated__` folder, i.e. the `css`/`cva` util, tokens, and in our case `jsx` components as well like `Box`.

Other libs/the app can then import from the styled-system lib:
```ts
import {Box, css} from "@your-repo/styled-system"
```
This respects the module boundaries as we're just exposing the public API we want, i.e. all the things generated by panda.

Note - to get this working (esp with storybook), we had to do some odd config things (see `libs/styled-system/panda.config.ts` and `apps/my-app/postcss.config.ts` and the `prepare` script in `package.json`). Here's a pretty simple `nx` repository with that setup: https://github.com/mayteio/panda-nx-next-storybook

**Reply by FranciscoKloganB** [reply]
Thank you so much @mayteio. I will try to migrate from `emotion` to `pandas`, since I have barelly any code written at this point. :)

I'll ping this post if I fail miserably. :dagger:

**Comment by tyteen4a03** [comment]
I'm using this in a monorepo but currently haven't moved the styled system into its own package yet. Does anybody have a solution for this scenario?

**Comment by SimeonC** [comment]
Thought I'd share my panda.config.mjs here as we found that there's a couple of things I found that make life easier.
* `emitPackage` places the generated code into node_modules, this seems to work more reliably than the tsconfig aliases and you don't need to `../../` on the outdir.
* `importMap` is the "back mapping"
* `workspaceRoot` I found was easiest pulling in from NX as you won't need to do any switching based on where it's running from
* `@local/styled-system` is just my own convention, you could use any name that doesn't collide with an existing npm package. The `outdir/importMap` just becomes the name used to import for this; `import { styled } from '@local/styled-system/jsx'`

```ts
import { defineConfig } from '@pandacss/dev';
import { workspaceRoot } from 'nx/src/utils/workspace-root';
import path from 'path';

export default defineConfig({
  include: [path.join(workspaceRoot, './apps/appName/src/**/*.{js,jsx,ts,tsx}')],
  importMap: '@local/styled-system',
  outdir: '@local/styled-system',
  emitPackage: true,
});
```
